### PR TITLE
Request messages from mailserver taking TTL into account

### LINKED ISSUE #6814: Public chats are missing messages on different devices
# Problem

Public chats are missing messages, that are not recovered even after multiple history requests.

I suspect that the problem is in the way we query history from mail servers. As I understood from the discussion in Status.im we do it in a following way: If at first I query envelopes from 13:10:11 till 13:15:17 and query was successful, next time i will start my query from 13:15:17 till some timestamp. The problem is that mail server can receive a message with timestamp lower than 13:15:17 after we sent back a history for 1st query.

# Implementation

Mail server uses (Expiry timestamp - TTL) as a timestamp for the message. Additionally there is small time window that is tolerated by any whisper node, afair it is 10s. I would suggest to adjust query with a common TTL and mentioned whisper window. 

It will increase rate of duplicates, but we will stop missing messages from mail server.

# Acceptance Criteria

Messages that are received by mail server after history was queried should be sent to the user on a subsequent queries.

link: https://github.com/status-im/status-react/issues/6757

### PR DESCRIPTION
**Pull request by pedropombeiro**
[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

fixes #6814

### Summary:

[comment]: # (Summarise the problem and how the pull request solves it)
This PR makes a slight change to the logic that requests messages from mailserver, so that the `from` value is adapted to take into account the message TTL and the max time drift allowed by a node, so `from` encompasses a bit more in the past.

### Review notes (optional):

I'm not sure if it's OK to expose `protocol/whisper-opts`, so would appreciate feedback on that. And of course, checking whether the logic was added to the most correct place.

status: ready

### CODE REVIEW

File: src/status_im/transport/message/protocol.cljs [hunk 1]
```diff
@@ -16,10 +16,11 @@
   (receive [this chat-id signature timestamp cofx] "Method producing all effects necessary for receiving the message record")
   (validate [this] "Method returning the message if it is valid or nil if it is not"))
 
-(def ^:private whisper-opts
-  {:ttl       10 ;; ttl of 10 sec
-   :powTarget config/pow-target
-   :powTime   config/pow-time})
+(def whisper-opts
+  {:whisper-drift-tolerance 10 ;; time drift that is tolerated by 10 seconds
```
Comment (rasom) [inline src/status_im/transport/message/protocol.cljs]: please put the comment on the previous line

File: src/status_im/mailserver/core.cljs [hunk 2]
```diff
@@ -292,6 +293,12 @@
        :from  from
        :to    to})))
 
+(defn adjust-request-for-transit-time
+  [{:keys [from] :as request}]
```
Comment (rasom) [inline src/status_im/mailserver/core.cljs]: `from` is not used

**Review by rasom — APPROVED** [review]

File: src/status_im/transport/message/protocol.cljs (hunk 1 above)
Comment (pedropombeiro) [reply src/status_im/transport/message/protocol.cljs]: Man, C#/Go habits die hard...

File: src/status_im/mailserver/core.cljs (hunk 2 above)
Comment (pedropombeiro) [reply src/status_im/mailserver/core.cljs]: I wasn't sure if there was some kind of convention to list keys we expected in an object, even if the value wasn't directly used. Removed.

File: src/status_im/mailserver/core.cljs [hunk 3]
```diff
@@ -292,6 +293,12 @@
        :from  from
        :to    to})))
 
+(defn adjust-request-for-transit-time
+  [request]
+  (let [ttl               (:ttl protocol/whisper-opts)
+        whisper-tolerance (:whisper-drift-tolerance protocol/whisper-opts)]
+    (update request :from #(- % (+ whisper-tolerance ttl)))))
```
Comment (cammellos) [inline src/status_im/mailserver/core.cljs]: this can be:
`(update request :from - (+ whisper-tolerance ttl))`

**Review by cammellos — APPROVED** [review]

File: src/status_im/transport/message/protocol.cljs (hunk 1 above)
Comment (rasom) [reply src/status_im/transport/message/protocol.cljs]: it actually can be used in clojure as well, but i encountered at least two cases when it was moved to the next line (after code re-formatting or incidentally) and so a comment was above the wrong line which is a bit misleading... And this specific thing can't happen if a comment was added before the line.

File: src/status_im/mailserver/core.cljs (hunk 3 above)
Comment (pedropombeiro) [reply src/status_im/mailserver/core.cljs]: good point!

**Comment by churik** [comment]
@PombeirP it is still WIP?
As I couldn't constantly reproduce the issue with missing messages, would be great to see any testing notes, if you can mention something more specific than fetching history for last 24 h or joining to chat with a lot (don't know exact value) of messages.

**Comment by pedropombeiro** [comment]
@churik updated status to ready. I think the easiest way to test this is to look at the logs (maybe with a test account and a single public chat) and check that when you bring the app to the foreground, it requests the messages with the expected compensated "from" value. I can add some additional logging showing both the usual and the compensated "from" value if you want.

**Comment by churik** [comment]
@PombeirP if I correctly understand, it should be changed comparing with current nightly?
I can compare two logs in this case.

**Comment by pedropombeiro** [comment]
@churik yeah the `from` value that gets sent to `requestMessages` should be smaller by 20 seconds than previously.

**Comment by churik** [comment]
@PombeirP 
1) In PR build the value `from` in `requestMessages`  is smaller by 20 seconds  than the time of request, i.e. 
```
t=2018-12-07T10:58:32+0100 lvl=info msg=RequestMessages                  package=status-go/services/sshext.PublicAPI request="{MailServerPeer:enode://7aa648d6e855950b2e3d3bf220c496e0cae4adfddef3e1e6062e6b177aec93bc6cdcf1282cb40d1656932ebfdd565729da440368d7c4da7dbd4d004b1ac02bf8@206.189.243.169:30504 From:1544090292 To:1544176712 Limit:0 Cursor: Topic:[0 0 0 0] Topics:[[145 119 79 246] [248 148 106 172]] SymKeyID:574eeb2accacc4e12f81deeb042608f255b827ac17addff18e054be600bdb535 Timeout:30ns}"
```
It is 10:58:12, so I suppose this test was passed. Previously this `from` was  the same as time of request (just to verify that I did all correctly)

**2) History for 24 h is not fetched. Waited for 5 + mins and didn't get message history.
3) Messages are not fetched from offline mode.**
Steps:
- join to any public channel
- turn off WI-FI
- send to this public chat message from another device
- go back to online
- wait for messages.
No new messages came.


Geth.log for both cases (tested on Mac OSx): 
[geth.log](https://github.com/status-im/status-react/files/2656820/geth.log)

**Comment by churik** [comment]
Not sure, but probably it can be related to https://notes.status.im/5apzxihzQcqRN0WzpvL6SQ 
will retest it again when problems will be resolved

<blockquote><div><strong><a href="https://notes.status.im/5apzxihzQcqRN0WzpvL6SQ">Suspicious mailserver traffic spike war-room - CodiMD</a></strong></div></blockquote>

**Comment by mandrigin** [comment]
@churik ping about re-testing this one ;-)

**Comment by churik** [comment]
@PombeirP 
Issues 2,3 are reproducible, so 
2) History for 24 h is not fetched. Waited for 5 + mins and didn't get message history.
Alert with fetching history is blinking (video: http://take.ms/Br6gdf)
**Steps:**
- create new account
- join to public chat


3) Messages are not fetched from offline mode.
**Steps:**

- join to any public channel
- turn off WI-FI
- send to this public chat message from another device
- go back to online
- wait for messages.
- No new messages came.
- Geth.log for both cases (tested on Mac OSx): [geth.log](https://github.com/status-im/status-react/files/2662540/geth.log)




<blockquote><div>Monosnap screenshot tool</div><div><strong><a href="https://monosnap.com/file/pQ7RkQ7leLdqkVMUV8YxmEyPDDuTyi">File "screencast 2018-12-10 10-19-33.mp4"</a></strong></div><div>Monosnap — the best tool to take & share your screenshots.</div></blockquote>

**Comment by pedropombeiro** [comment]
@churik I'm wondering if there are any other steps required. If I follow the steps you mentioned, I get the message that was posted while I was offline. I've tried waiting 1 minute while in airplane mode, and also with just WiFi off, both times it arrived as expected. I'm on a rebased branch, maybe something else has been fixed in develop in the meantime?

**Comment by churik** [comment]
@PombeirP 
Can you rebase it to current develop and send it to test again?
I tested on  https://github.com/status-im/status-react/pull/6988#issuecomment-444825509 , so probably rebase can help.

**Comment by pedropombeiro** [comment]
@churik Jenkins is being even more of a lottery than usual today, but hopefully the build will pass soon.

**Comment by pedropombeiro** [comment]
@churik There seems to be something wrong with the Windows build, but all the others passed, so if you don't mind, please get the artifact you need to test at https://ci.status.im/blue/organizations/jenkins/status-react%2Fpull%20requests/detail/PR-6988/7/pipeline. Thanks

<blockquote><div><strong><a href="https://ci.status.im/blue/organizations/jenkins/status-react%2Fpull%20requests/detail/PR-6988/7/pipeline">Jenkins</a></strong></div></blockquote>

**Comment by churik** [comment]
@PombeirP tried on build from https://github.com/status-im/status-react/pull/6988#issuecomment-445921843 - nothing is fetched when I join to #status chat. 
Tried from Mac OS and IOS, from upgraded build and after freshly installation - history is not fetched for me.
[geth.log](https://github.com/status-im/status-react/files/2666773/geth.log)

**Comment by pedropombeiro** [comment]
After discussing with @cammellos, I think it's best to close this PR in favor of moving the mailserver fetching logic to the status-go as part of a refactor that was already being discussed.

**Pull request closed without merging**